File: problems/167-two-sum-ii-input-array-is-sorted/two-pointers.js
/**
 * 排序 + 双指针：O(NlogN)
 */
var twoSum = function(nums, targetSum) {
  const objects = nums.map((value, index) => ({
    index: index + 1,
    value
  })).sort((a, b) => a.value - b.value)

  let [L, R] = [0, objects.length - 1];
  while (L < R) {
    const sum = objects[L].value + objects[R].value
    if (sum > targetSum) {
      --R
    } else if (sum < targetSum) {
      ++L
    } else {
      return [objects[L].index, objects[R].index]
    }
  }
  return [-1, -1]
};

File: problems/167-two-sum-ii-input-array-is-sorted/two-pointers.cpp
/**
 * 双指针
 * 
 * 时间：`O(N)`, 0ms
 */

class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int L = 0, R = numbers.size() - 1;
        while (L < R) {
            int sum = numbers[L] + numbers[R];
            if (sum == target) {
                return {L + 1, R + 1};
                ++L;
                --R;
            } else if (sum < target) {
                ++L;
            } else {
                --R;
            }
        }
        return {};
    }
};